sw/device/lib/dif/README.md

```diff
@@ -12,3 +12,56 @@ libraries.
 There is one DIF library per hardware IP, and each one contains the DIFs
 required to actuate all of the specification-required functionality of the
 hardware they are written for.
+
+## DIF Style Guide
+
+DIFs are very low-level software, so they have a more rigorous coding style than
+other parts of the codebase.
+
+DIFs should follow the [OpenTitan C/C++ style
+guide](https://docs.opentitan.org/doc/rm/c_cpp_coding_style/) where it does not
+contradict with the guidelines below.
+
+The guidelines below apply to writing DIFs, and code should be written in a
+similar style to the existing DIF libraries in this directory.
+
+### DIF Library Guidance
+
+* DIF libraries must be written in C.
+* DIF libraries can only depend on the headers and libraries in the
+  `sw/device/lib/base` directory.
+* DIF libraries must not depend on other DIF libraries. Exercising DIF
+  functionality may require an environment set up using another DIF library, but
+  DIFs must not call DIFs in other DIF libraries.
+
+### DIF Guidance
+
+* DIF declarations must match their definitions exactly.
+  * Scalar arguments should not be marked `const` in DIF signatures.
+
+* DIFs must use enum return codes rather than booleans for reporting errors. If
+  a DIF can either error or produce a value, it should return an enum return
+  code, and use an out-parameter for returning the produced value.
+
+* DIFs must check their arguments against preconditions using "guard
+  statements". A guard statement is a simple if statement at the start of a
+  function which only returns an error code if the preconditions are not met.
+  Guard statements must cover the following checks:
+  * DIFs must ensure their pointer arguments are non-null, unless that pointer
+    is for an optional out-parameter.
+  * DIFs must ensure, if they only accept a subset of an enum, that the argument
+    is within that subset. However, DIFs may assume that the argument is one of
+    the enum variants.
+  * DIFs must not have side-effects before any guard statements. Side-effects
+    include (but are not limited to) writing to memory, including memory-mapped
+    hardware, and modifying processor CSRs.
+
+* Switch statements in DIFs must always have a default case, including when
+  switching on an enum value (an "enum switch").
+  * The default case of an enum switch should report an error with values that
+    are not an enum variant.
+  * Enum switches do not need a `case` for enum variants that are unreachable
+    due to a guard statement.
+
+* DIFs must use `sw/device/lib/mmio.h` for accessing memory-mapped hardware.
+  DIFs must not use `sw/device/lib/memory.h` for accessing memory-mapped hardware.
```

Comment:
Something something "Since this would require casting away `volatile`, this is already considered Undefined Behavior by C."

Reply:
I think it's not worth opening the implementation details, given that we have provided One True Way of accessing MMIO regions already.